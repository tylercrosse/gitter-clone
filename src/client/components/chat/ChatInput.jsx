import React    from 'react';
import { Link } from 'react-router';

class ChatInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      draft: ''
    };
  }
  updateDraft(e) {
    this.setState({
      draft: e.target.value
    });
  }
  handleSubmit(e) {
    e.preventDefault();
    this.props.onMessageSubmit({
      text: this.state.draft,
      username: this.props.user.username
    });
    this.setState({
      draft: ''
    });
  }
  render() {
    const chatAuth = this.props.user.loggedIn ? (
      <div className="chat-input-container">
        <div className="chat-input-avatar">
          <img
            className="avatar"
            src={'http://i.pravatar.cc/30?u=' + this.props.user.username} alt={this.props.user.username}
          />
        </div>
        <form onSubmit={(e) => this.handleSubmit(e)}>
          <textarea
            onChange={(e) => this.updateDraft(e)}
            value={this.state.draft}
            placeholder="Click here to type a chat message."
            name="chat"
            type="text"
            autoFocus
            autoComplete="off"
          />
          <button type="submit">Send</button>
        </form>
      </div>
    ) : (
      <div className="chat-input-container">
        <Link className="chat-input-btn" to="/">Sign in to start talking</Link>
      </div>
    );
    return (
      <footer className="chat-input">
        {chatAuth}
      </footer>
    );
  }
}

export default ChatInput;
