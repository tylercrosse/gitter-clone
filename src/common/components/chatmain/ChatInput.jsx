import React, { PropTypes } from 'react';
import debounce from 'lodash/debounce';

const DEBOUNCE_TIME = 3000;
const TIME_CUSHION = 200;

class ChatInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      draft: ''
    };
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    // NOTE the debounce time for addTypingUser needs to be slightly less to avoid race condition-esque situation with the server socket responses in the wrong order
    this.addTypingUser = debounce(
      this.props.addTypingUser,
      DEBOUNCE_TIME - TIME_CUSHION,
      { leading: true }
    );
    this.removeTypingUser = debounce(this.props.removeTypingUser, DEBOUNCE_TIME);
  }
  handleKeyPress(e) {
    if (e.key === 'Enter' && !e.nativeEvent.shiftKey) {
      this.handleSubmit(e);
    }
  }
  handleChange(e) {
    this.setState({
      draft: e.target.value
    });
    this.addTypingUser({
      username: this.props.user.name
    });
    this.removeTypingUser({
      username: this.props.user.name
    });
  }
  handleSubmit(e) {
    e.preventDefault();
    if (!this.state.draft.trim()) return;
    const convo = this.props.routeParams.convo;
    this.props.onMessageSubmit({
      userId: this.props.user.id,
      text: this.state.draft,
      convo
    });
    this.props.removeTypingUser({
      username: this.props.user.name
    });
    this.setState({
      draft: ''
    });
  }
  render() {
    let inputContent;
    if (this.props.user.loggedIn) {
      inputContent = (
        <div className="chat-input-container">
          <div className="chat-input-avatar">
            <img
              className="avatar"
              src={'http://i.pravatar.cc/30?u=' + this.props.user.name}
              alt={this.props.user.name}
            />
          </div>
          <form onSubmit={this.handleSubmit}>
            <textarea
              onKeyPress={this.handleKeyPress}
              onChange={this.handleChange}
              value={this.state.draft}
              placeholder="Click here to type a chat message. Supports Github flavoured markdown."
              name="chat"
              type="text"
              autoFocus
              autoComplete="off"
            />
            <button type="submit">Send</button>
          </form>
        </div>
      );
    } else {
      inputContent = (
        <div className="chat-input-container">
          <button
            onClick={this.props.openSignInModal}
            className="chat-input-btn"
          >
            Sign in to start talking
          </button>
        </div>
      );
    }
    return (
      <footer className="chat-input">
        {inputContent}
      </footer>
    );
  }
}

ChatInput.propTypes = {
  addTypingUser: PropTypes.func.isRequired,
  removeTypingUser: PropTypes.func.isRequired,
  onMessageSubmit: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  routeParams: PropTypes.object.isRequired
};

export default ChatInput;
