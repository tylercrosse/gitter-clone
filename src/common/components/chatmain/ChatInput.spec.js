import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ChatInput from './ChatInput';

const setup = (propOverrides) => {
  const props = Object.assign({
    user: {
      name: 'Bob',
      loggedIn: true
    },
    routeParams: {
      convo: 'redux'
    },
    onMessageSubmit: jest.fn(),
    addTypingUser: jest.fn(),
    removeTypingUser: jest.fn(),
    openSignInModal: jest.fn()
  }, propOverrides);

  const component = <ChatInput {...props} />;
  const wrapper = shallow(component);

  return {
    props,
    component,
    wrapper
  };
};

describe('<ChatInput />', () => {
  describe('logged in', () => {
    it('should render correctly', () => {
      const { wrapper } = setup();
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should contain <form />', () => {
      const { wrapper } = setup();
      expect(wrapper.find('form'))
        .toHaveLength(1);
    });

    it('should simulate textarea field', () => {
      const value = 'My new value';
      const { wrapper } = setup();
      wrapper.find('textarea').simulate('change', {target: {value}});
      expect(wrapper.find('textarea').props().value)
        .toEqual(value);
    });

    it('should submit on enter, but not shift+enter', () => {
      const shiftEnterEvent = {
        key: 'Enter',
        nativeEvent: {shiftKey: true}
      };
      const enterEvent = {
        key: 'Enter',
        nativeEvent: {shiftKey: false}
      };
      const { component } = setup();
      const wrapper = mount(component);
      wrapper.instance().handleSubmit = jest.fn();
      wrapper.instance().handleKeyPress(shiftEnterEvent);
      expect(wrapper.instance().handleSubmit)
        .toHaveBeenCalledTimes(0);
      wrapper.instance().handleKeyPress(enterEvent);
      expect(wrapper.instance().handleSubmit)
        .toHaveBeenCalledTimes(1);
    });

    it('should call submit action on submit with text', () => {
      const value = 'My new value';
      const { props, component } = setup();
      const wrapper = mount(component);
      // no message (trim conditional)
      wrapper.find('form').simulate('submit');
      expect(props.onMessageSubmit)
        .toHaveBeenCalledTimes(0);
      // with message
      wrapper.find('textarea').simulate('change', {target: {value}});
      wrapper.find('form').simulate('submit');
      expect(props.onMessageSubmit)
        .toHaveBeenCalledWith(expect.objectContaining({text: value}));
    });
  });

  describe('logged out', () => {
    it('should render correctly', () => {
      const { wrapper } = setup({user: {loggedIn: false}});
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not contain <form />', () => {
      const { wrapper } = setup({user: {loggedIn: false}});
      expect(wrapper.find('form'))
        .toHaveLength(0);
    });
  });
});
