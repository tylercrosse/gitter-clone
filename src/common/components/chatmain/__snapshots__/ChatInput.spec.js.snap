// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatInput /> logged in should render correctly 1`] = `
<footer
  className="chat-input"
>
  <div
    className="chat-input-container"
  >
    <div
      className="chat-input-avatar"
    >
      <img
        alt="Bob"
        className="avatar"
        src="http://i.pravatar.cc/30?u=Bob"
      />
    </div>
    <form
      onSubmit={[Function]}
    >
      <textarea
        autoComplete="off"
        autoFocus={true}
        name="chat"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Click here to type a chat message. Supports Github flavoured markdown."
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</footer>
`;

exports[`<ChatInput /> logged out should render correctly 1`] = `
<footer
  className="chat-input"
>
  <div
    className="chat-input-container"
  >
    <button
      className="chat-input-btn"
      onClick={[Function]}
    >
      Sign in to start talking
    </button>
  </div>
</footer>
`;
